### ISSUE TITLE
Momentarily pause while dragging the window under system load

### ISSUE THREAD

**Issue by marioroy** [xanmod/linux#402]
XanMod LTS 6.1.53
Including XanMod RT 6.1.46 

I'm hoping to complete the [XanMod kernel on Clear Linux](https://community.clearlinux.org/t/running-xanmod-kernel-on-clear-linux/9131) project this week. This is one stress test where the XanMod kernels struggle.

1. Build [primesieve](https://github.com/kimwalisch/primesieve) from source, so that it builds with AVX2 or higher vector instructions.
2. Launch Google Chrome or any Chromium-based browser. A blank page is fine.
3. In a terminal window, run `primesieve 1e13`. Pressing `Ctrl-C` will stop it.
4. Try moving the Chrome window around the desktop in a circle. Does the movement momentarily pause?

The Clear Linux LTS kernel 6.1.53 passes. No pause and smooth.

Running `chrt -r 20 /opt/google/chrome/google-chrome` works. No pause moving the window while the system is under 100% compute load.

**Comment by marioroy** [comment]
<details>
<summary>Spec file for XanMod LTS 6.1.53</summary>

```text
#
# note to self: Linus releases need to be named 5.x.0 not 5.x or various
# things break
#
#

Name:           linux-xmlts
Version:        6.1.53
Release:        2
License:        GPL-2.0
Summary:        The Linux kernel
Url:            http://www.kernel.org/
Group:          kernel
Source0:        https://github.com/xanmod/linux/archive/refs/tags/6.1.53-xanmod1.tar.gz
Source1:        config
Source2:        cmdline

%define ktarget  xmlts
%define kversion %{version}-%{release}.%{ktarget}
%define xmcustomver 1

#BuildRequires:  buildreq-kernel

Requires: systemd-bin
Requires: init-rdahead-extras
Requires: linux-xmlts-license = %{version}-%{release}

# don't strip .ko files!
%global __os_install_post %{nil}
%define debug_package %{nil}
%define __strip /bin/true

#cve.start cve patches from 0001 to 050
#cve.end

#mainline: Mainline patches, upstream backport and fixes from 0051 to 0099
#mainline.end

#Serie.clr 01XX: Clear Linux patches
Patch0101: 0101-i8042-decrease-debug-message-level-to-info.patch
Patch0102: 0102-increase-the-ext4-default-commit-age.patch
Patch0104: 0104-pci-pme-wakeups.patch
Patch0105: 0105-ksm-wakeups.patch
Patch0106: 0106-intel_idle-tweak-cpuidle-cstates.patch
Patch0107: 0107-bootstats-add-printk-s-to-measure-boot-time-in-more-.patch
Patch0108: 0108-smpboot-reuse-timer-calibration.patch
Patch0111: 0111-ipv4-tcp-allow-the-memory-tuning-for-tcp-to-go-a-lit.patch
Patch0114: 0114-add-boot-option-to-allow-unsigned-modules.patch
Patch0116: 0116-migrate-some-systemd-defaults-to-the-kernel-defaults.patch
Patch0117: 0117-xattr-allow-setting-user.-attributes-on-symlinks-by-.patch
Patch0122: 0122-ata-libahci-ignore-staggered-spin-up.patch
Patch0123: 0123-print-CPU-that-faults.patch
Patch0124: 0124-x86-microcode-Add-an-option-to-reload-microcode-even.patch
Patch0125: 0125-nvme-workaround.patch
Patch0126: 0126-don-t-report-an-error-if-PowerClamp-run-on-other-CPU.patch
Patch0127: 0127-lib-raid6-add-patch.patch
Patch0128: 0128-itmt_epb-use-epb-to-scale-itmt.patch
Patch0130: 0130-itmt2-ADL-fixes.patch
Patch0131: 0131-add-a-per-cpu-minimum-high-watermark-an-tune-batch-s.patch
Patch0134: 0134-md-raid6-algorithms-scale-test-duration-for-speedier.patch
Patch0135: 0135-initcall-only-print-non-zero-initcall-debug-to-speed.patch
Patch0136: scale.patch
Patch0137: libsgrowdown.patch
Patch0138: kdf-boottime.patch
Patch0139: adlrdt.patch
Patch0140: kvm-printk.patch
Patch0141: epp-retune.patch
Patch0142: tcptuning.patch
Patch0143: 0001-powerbump-functionality.patch
Patch0144: 0002-add-networking-support-for-powerbump.patch
Patch0145: 0003-futex-bump.patch
Patch0146: 0001-add-umonitor-umwait-C0.x-C-states.patch
Patch0147: 0001-mm-memcontrol-add-some-branch-hints-based-on-gcov-an.patch
Patch0148: 0002-sched-core-add-some-branch-hints-based-on-gcov-analy.patch
Patch0149: netscale.patch
Patch0162: 0162-xm-extra-optmization-flags.patch
#Serie.end

%description
The Linux kernel.

%package extra
License:        GPL-2.0
Summary:        The Linux kernel extra files
Group:          kernel
Requires:       linux-xmlts-license = %{version}-%{release}

%description extra
Linux kernel extra files

%package license
Summary: license components for the linux package.
Group: Default

%description license
license components for the linux package.

%package cpio
License:        GPL-2.0
Summary:        cpio file with kenrel modules
Group:          kernel

%description cpio
Creates a cpio file with some modules

%package dev
License:        GPL-2.0
Summary:        The Linux kernel
Group:          kernel
Requires:       linux-xmlts = %{version}-%{release}
Requires:       linux-xmlts-extra = %{version}-%{release}
Requires:       linux-xmlts-license = %{version}-%{release}

%description dev
Linux kernel build files

%prep
%setup -q -n linux-%{version}-xanmod%{xmcustomver}

#cve.patch.start cve patches
#cve.patch.end

#mainline.patch.start Mainline patches, upstream backport and fixes
#mainline.patch.end

#Serie.patch.start Clear Linux patches
%patch -P 101 -p1
%patch -P 102 -p1
%patch -P 104 -p1
%patch -P 105 -p1
%patch -P 106 -p1
%patch -P 107 -p1
%patch -P 108 -p1
%patch -P 111 -p1
%patch -P 114 -p1
%patch -P 116 -p1
%patch -P 117 -p1
%patch -P 122 -p1
%patch -P 123 -p1
%patch -P 124 -p1
%patch -P 125 -p1
%patch -P 126 -p1
%patch -P 127 -p1
%patch -P 128 -p1
%patch -P 130 -p1
%patch -P 131 -p1
%patch -P 134 -p1
%patch -P 135 -p1
%patch -P 136 -p1
%patch -P 137 -p1
%patch -P 138 -p1
%patch -P 139 -p1
%patch -P 140 -p1
%patch -P 141 -p1
%patch -P 142 -p1
%patch -P 143 -p1
%patch -P 144 -p1
%patch -P 145 -p1
%patch -P 146 -p1
%patch -P 147 -p1
%patch -P 148 -p1
%patch -P 149 -p1
%patch -P 162 -p1
#Serie.patch.end


cp %{SOURCE1} .config

# Set v3 CPU (x86-64-v3).
scripts/config -e GENERIC_CPU3

# Disable DEBUG.
%if 1
scripts/config -d DEBUG_INFO
scripts/config -d DEBUG_INFO_BTF
scripts/config -d DEBUG_INFO_DWARF4
scripts/config -d DEBUG_INFO_DWARF5
scripts/config -d PAHOLE_HAS_SPLIT_BTF
scripts/config -d DEBUG_INFO_BTF_MODULES
scripts/config -d SLUB_DEBUG
scripts/config -d PM_DEBUG
scripts/config -d PM_ADVANCED_DEBUG
scripts/config -d PM_SLEEP_DEBUG
scripts/config -d ACPI_DEBUG
scripts/config -d SCHED_DEBUG
scripts/config -d LATENCYTOP
scripts/config -d DEBUG_PREEMPT
%endif

# Enable full tickless mode.
# https://docs.kernel.org/timers/no_hz.html
%if 1
scripts/config -d HZ_PERIODIC
scripts/config -d NO_HZ_IDLE
scripts/config -d CONTEXT_TRACKING_FORCE
scripts/config -d CONTEXT_TRACKING_USER_FORCE
scripts/config -e NO_HZ_FULL_NODEF
scripts/config -e NO_HZ_FULL
scripts/config -e NO_HZ
scripts/config -e NO_HZ_COMMON
scripts/config -e CONTEXT_TRACKING
%endif

# Set RCU options.
# https://www.kernel.org/doc/Documentation/RCU/Design/Data-Structures/Data-Structures.html
# https://raw.githubusercontent.com/CachyOS/linux-cachyos/master/linux-bore/config
# https://linuxmusicians.com/viewtopic.php?t=13336 # RCU_BOOST_PRIO, RCU_BOOST_DELAY
# Reduce to 98 the maximum real time priority if defined in limits.conf
# https://forums.gentoo.org/viewtopic-p-6916148.html
%if 1
scripts/config -e RCU_EXPERT
scripts/config -d FORCE_TASKS_RCU
scripts/config -d FORCE_TASKS_RUDE_RCU
scripts/config -d FORCE_TASKS_TRACE_RCU
scripts/config --set-val RCU_FANOUT 64
scripts/config --set-val RCU_FANOUT_LEAF 16
scripts/config -e RCU_BOOST
scripts/config --set-val RCU_BOOST_PRIO 99
scripts/config --set-val RCU_BOOST_DELAY 331
scripts/config -d RCU_EXP_KTHREAD
scripts/config -e RCU_NOCB_CPU
scripts/config -e RCU_NOCB_CPU_DEFAULT_ALL
scripts/config -d RCU_NOCB_CPU_CB_BOOST
scripts/config -d TASKS_TRACE_RCU_READ_MB
scripts/config -d RCU_LAZY
%endif

# Enable full preempt.
# https://raw.githubusercontent.com/CachyOS/linux-cachyos/master/linux-bore/config
%if 1
scripts/config -e PREEMPT_BUILD
scripts/config -d PREEMPT_NONE
scripts/config -d PREEMPT_VOLUNTARY
scripts/config -e PREEMPT
scripts/config -e PREEMPT_COUNT
scripts/config -e PREEMPTION
scripts/config -e PREEMPT_DYNAMIC
%endif

mv .config config

%build
BuildKernel() {

    Target=$1
    Arch=x86_64
    ExtraVer="-%{release}.${Target}"

    rm -f localversion

    perl -p -i -e "s/^EXTRAVERSION.*/EXTRAVERSION = ${ExtraVer}/" Makefile

    make O=${Target} -s mrproper
    cp config ${Target}/.config

    make O=${Target} -s ARCH=${Arch} olddefconfig
    make O=${Target} -s ARCH=${Arch} CONFIG_DEBUG_SECTION_MISMATCH=y %{?_smp_mflags} %{?sparse_mflags}
}

BuildKernel %{ktarget}

%install

InstallKernel() {

    Target=$1
    Kversion=$2
    Arch=x86_64
    KernelDir=%{buildroot}/usr/lib/kernel
    DevDir=%{buildroot}/usr/lib/modules/${Kversion}/build

    mkdir   -p ${KernelDir}
    install -m 644 ${Target}/.config    ${KernelDir}/config-${Kversion}
    install -m 644 ${Target}/System.map ${KernelDir}/System.map-${Kversion}
    install -m 644 ${Target}/vmlinux    ${KernelDir}/vmlinux-${Kversion}
    install -m 644 %{SOURCE2}           ${KernelDir}/cmdline-${Kversion}
    cp  ${Target}/arch/x86/boot/bzImage ${KernelDir}/org.clearlinux.${Target}.%{version}-%{release}
    chmod 755 ${KernelDir}/org.clearlinux.${Target}.%{version}-%{release}

    mkdir -p %{buildroot}/usr/lib/modules
    make O=${Target} -s ARCH=${Arch} INSTALL_MOD_PATH=%{buildroot}/usr modules_install

    rm -f %{buildroot}/usr/lib/modules/${Kversion}/build
    rm -f %{buildroot}/usr/lib/modules/${Kversion}/source

    mkdir -p ${DevDir}
    find . -type f -a '(' -name 'Makefile*' -o -name 'Kbuild*' -o -name 'Kconfig*' ')' -exec cp -t ${DevDir} --parents -pr {} +
    find . -type f -a '(' -name '*.sh' -o -name '*.pl' ')' -exec cp -t ${DevDir} --parents -pr {} +
    cp -t ${DevDir} -pr ${Target}/{Module.symvers,tools}
    ln -s ../../../kernel/config-${Kversion} ${DevDir}/.config
    ln -s ../../../kernel/System.map-${Kversion} ${DevDir}/System.map
    cp -t ${DevDir} --parents -pr arch/x86/include
    cp -t ${DevDir}/arch/x86/include -pr ${Target}/arch/x86/include/*
    cp -t ${DevDir}/include -pr include/*
    cp -t ${DevDir}/include -pr ${Target}/include/*
    cp -t ${DevDir} --parents -pr scripts/*
    cp -t ${DevDir}/scripts -pr ${Target}/scripts/*
    find  ${DevDir}/scripts -type f -name '*.[cho]' -exec rm -v {} +
    find  ${DevDir} -type f -name '*.cmd' -exec rm -v {} +
    # Cleanup any dangling links
    find ${DevDir} -type l -follow -exec rm -v {} +

    # Kernel default target link
    ln -s org.clearlinux.${Target}.%{version}-%{release} %{buildroot}/usr/lib/kernel/default-${Target}
}

# cpio file for keyboard drivers
createCPIO() {

    Target=$1
    Kversion=$2
    KernelDir=%{buildroot}/usr/lib/kernel
    ModDir=/usr/lib/modules/${Kversion}

    mkdir -p cpiofile${ModDir}/kernel/drivers/input/{serio,keyboard}
    mkdir -p cpiofile${ModDir}/kernel/drivers/hid
    cp %{buildroot}${ModDir}/kernel/drivers/input/serio/i8042.ko      cpiofile${ModDir}/kernel/drivers/input/serio
    cp %{buildroot}${ModDir}/kernel/drivers/input/serio/libps2.ko     cpiofile${ModDir}/kernel/drivers/input/serio
    cp %{buildroot}${ModDir}/kernel/drivers/input/keyboard/atkbd.ko   cpiofile${ModDir}/kernel/drivers/input/keyboard
    cp %{buildroot}${ModDir}/kernel/drivers/hid/hid-logitech-dj.ko    cpiofile${ModDir}/kernel/drivers/hid
    cp %{buildroot}${ModDir}/kernel/drivers/hid/hid-logitech-hidpp.ko cpiofile${ModDir}/kernel/drivers/hid
    cp %{buildroot}${ModDir}/kernel/drivers/hid/hid-apple.ko          cpiofile${ModDir}/kernel/drivers/hid
    cp %{buildroot}${ModDir}/modules.order   cpiofile${ModDir}
    cp %{buildroot}${ModDir}/modules.builtin cpiofile${ModDir}

    depmod -b cpiofile/usr ${Kversion}

    (
      cd cpiofile
      find . | cpio --create --format=newc \
        | xz --check=crc32 --lzma2=dict=512KiB > ${KernelDir}/initrd-org.clearlinux.${Target}.%{version}-%{release}
    )
}

InstallKernel %{ktarget} %{kversion}

createCPIO %{ktarget} %{kversion}

rm -rf %{buildroot}/usr/lib/firmware

mkdir -p %{buildroot}/usr/share/package-licenses/linux-xmlts
cp COPYING %{buildroot}/usr/share/package-licenses/linux-xmlts/COPYING
cp -a LICENSES/* %{buildroot}/usr/share/package-licenses/linux-xmlts

%postun
rm -fr /var/lib/dkms/*/kernel-%{kversion}-x86_64
rm -fr /var/lib/dkms/*/*/%{kversion}

%files
%dir /usr/lib/kernel
%dir /usr/lib/modules/%{kversion}
/usr/lib/kernel/config-%{kversion}
/usr/lib/kernel/cmdline-%{kversion}
/usr/lib/kernel/org.clearlinux.%{ktarget}.%{version}-%{release}
/usr/lib/kernel/default-%{ktarget}
/usr/lib/modules/%{kversion}/kernel
/usr/lib/modules/%{kversion}/modules.*

%files extra
%dir /usr/lib/kernel
/usr/lib/kernel/System.map-%{kversion}
/usr/lib/kernel/vmlinux-%{kversion}

%files license
%defattr(0644,root,root,0755)
/usr/share/package-licenses/linux-xmlts

%files cpio
/usr/lib/kernel/initrd-org.clearlinux.%{ktarget}.%{version}-%{release}

%files dev
%defattr(-,root,root)
/usr/lib/modules/%{kversion}/build

```

</details>

**Comment by marioroy** [comment]
<details>
<summary>Spec file for XanMod RT 6.1.46</summary>

```text
#
# This is the kernel native kernel plus the Preempt-RT patches
#
#
#

Name:           linux-xmrt
Version:        6.1.46
Release:        2
License:        GPL-2.0
Summary:        The Linux kernel with Preempt-RT
Url:            https://www.kernel.org
Group:          kernel
Source0:        https://github.com/xanmod/linux/archive/refs/tags/6.1.46-rt13-xanmod1.tar.gz
Source1:        config
Source2:        cmdline

%define ktarget  xmrt
%define kversion %{version}-%{release}.%{ktarget}
%define xmcustomver 1
%define xmcustomver_rt 13

#BuildRequires:  buildreq-kernel

Requires: systemd-bin
Requires: init-rdahead-extras
Requires: linux-xmrt-license = %{version}-%{release}

# don't strip .ko files!
%global __os_install_post %{nil}
%define debug_package %{nil}
%define __strip /bin/true

#cve.start cve patches from 0001 to 050
#cve.end

#mainline: Mainline patches, upstream backport and fixes from 0051 to 0099
#mainline.end

#Serie.clr 01XX: Clear Linux patches
Patch0101: 0101-i8042-decrease-debug-message-level-to-info.patch
Patch0102: 0102-Increase-the-ext4-default-commit-age.patch
Patch0104: 0104-pci-pme-wakeups.patch
Patch0105: 0105-ksm-wakeups.patch
Patch0106: 0106-intel_idle-tweak-cpuidle-cstates.patch
Patch0107: 0107-bootstats-add-printk-s-to-measure-boot-time-in-more-.patch
Patch0108: 0108-smpboot-reuse-timer-calibration.patch
Patch0111: 0111-ipv4-tcp-allow-the-memory-tuning-for-tcp-to-go-a-lit.patch
Patch0115: 0115-Add-boot-option-to-allow-unsigned-modules.patch
Patch0117: 0117-Migrate-some-systemd-defaults-to-the-kernel-defaults.patch
Patch0118: 0118-xattr-allow-setting-user.-attributes-on-symlinks-by-.patch
Patch0122: 0122-zero-extra-registers.patch
Patch0124: 0124-ata-libahci-ignore-staggered-spin-up.patch
Patch0125: 0125-print-CPU-that-faults.patch
Patch0126: 0126-x86-microcode-Force-update-a-uCode-even-if-the-rev-i.patch
Patch0128: 0128-fix-bug-in-ucode-force-reload-revision-check.patch
Patch0129: 0129-nvme-workaround.patch
Patch0130: 0130-Don-t-report-an-error-if-PowerClamp-run-on-other-CPU.patch
Patch0131: 0131-overload-on-wakeup.patch
Patch0162: 0162-xm-extra-optmization-flags.patch
#Serie.end

%description
The Linux kernel.

%package extra
License:        GPL-2.0
Summary:        The Linux kernel extra files
Group:          kernel
Requires:       linux-xmrt-license = %{version}-%{release}

%description extra
Linux kernel extra files

%package license
Summary: license components for the linux package.
Group: Default

%description license
license components for the linux package.

%package cpio
License:        GPL-2.0
Summary:        cpio file with kenrel modules
Group:          kernel

%description cpio
Creates a cpio file with some modules

%package dev
License:        GPL-2.0
Summary:        The Linux kernel
Group:          kernel
Requires:       linux-xmrt = %{version}-%{release}
Requires:       linux-xmrt-extra = %{version}-%{release}
Requires:       linux-xmrt-license = %{version}-%{release}

%description dev
Linux kernel build files

%prep
%setup -q -n linux-%{version}-rt%{xmcustomver_rt}-xanmod%{xmcustomver}

#cve.patch.start cve patches
#cve.patch.end

#mainline.patch.start Mainline patches, upstream backport and fixes
#mainline.patch.end

#Serie.patch.start Clear Linux patches
%patch -P 101 -p1
%patch -P 102 -p1
%patch -P 104 -p1
%patch -P 105 -p1
%patch -P 106 -p1
%patch -P 107 -p1
%patch -P 108 -p1
%patch -P 111 -p1
%patch -P 115 -p1
%patch -P 117 -p1
%patch -P 118 -p1
%patch -P 122 -p1
%patch -P 124 -p1
%patch -P 125 -p1
%patch -P 126 -p1
%patch -P 128 -p1
%patch -P 129 -p1
%patch -P 130 -p1
%patch -P 131 -p1
%patch -P 162 -p1
#Serie.patch.end


cp %{SOURCE1} .config

# Set v3 CPU (x86-64-v3).
scripts/config -e GENERIC_CPU3

# Disable DEBUG.
%if 1
scripts/config -d DEBUG_INFO
scripts/config -d DEBUG_INFO_BTF
scripts/config -d DEBUG_INFO_DWARF4
scripts/config -d DEBUG_INFO_DWARF5
scripts/config -d PAHOLE_HAS_SPLIT_BTF
scripts/config -d DEBUG_INFO_BTF_MODULES
scripts/config -d SLUB_DEBUG
scripts/config -d PM_DEBUG
scripts/config -d PM_ADVANCED_DEBUG
scripts/config -d PM_SLEEP_DEBUG
scripts/config -d ACPI_DEBUG
scripts/config -d SCHED_DEBUG
scripts/config -d LATENCYTOP
scripts/config -d DEBUG_PREEMPT
%endif

# Enable full tickless mode.
# https://docs.kernel.org/timers/no_hz.html
%if 1
scripts/config -d HZ_PERIODIC
scripts/config -d NO_HZ_IDLE
scripts/config -d CONTEXT_TRACKING_FORCE
scripts/config -d CONTEXT_TRACKING_USER_FORCE
scripts/config -e NO_HZ_FULL_NODEF
scripts/config -e NO_HZ_FULL
scripts/config -e NO_HZ
scripts/config -e NO_HZ_COMMON
scripts/config -e CONTEXT_TRACKING
%endif

# Set RCU options.
# https://www.kernel.org/doc/Documentation/RCU/Design/Data-Structures/Data-Structures.html
# https://raw.githubusercontent.com/CachyOS/linux-cachyos/master/linux-bore/config
# https://linuxmusicians.com/viewtopic.php?t=13336 # RCU_BOOST_PRIO, RCU_BOOST_DELAY
# Reduce to 98 the maximum real time priority if defined in limits.conf
# https://forums.gentoo.org/viewtopic-p-6916148.html
%if 1
scripts/config -e RCU_EXPERT
scripts/config -d FORCE_TASKS_RCU
scripts/config -d FORCE_TASKS_RUDE_RCU
scripts/config -d FORCE_TASKS_TRACE_RCU
scripts/config --set-val RCU_FANOUT 64
scripts/config --set-val RCU_FANOUT_LEAF 16
scripts/config -e RCU_BOOST
scripts/config --set-val RCU_BOOST_PRIO 99
scripts/config --set-val RCU_BOOST_DELAY 331
scripts/config -d RCU_EXP_KTHREAD
scripts/config -e RCU_NOCB_CPU
scripts/config -e RCU_NOCB_CPU_DEFAULT_ALL
scripts/config -d RCU_NOCB_CPU_CB_BOOST
scripts/config -d TASKS_TRACE_RCU_READ_MB
scripts/config -d RCU_LAZY
%endif

# Disable transparent huge pages
# https://rigtorp.se/low-latency-guide/
%if 1
scripts/config -d TRANSPARENT_HUGEPAGE
scripts/config -d TRANSPARENT_HUGEPAGE_ALWAYS
%endif

# Enable full preempt-rt.
%if 1
scripts/config -e PREEMPT_BUILD
scripts/config -d PREEMPT_NONE
scripts/config -d PREEMPT_VOLUNTARY
scripts/config -d PREEMPT
scripts/config -e HAVE_PREEMPT_LAZY
scripts/config -e PREEMPT_LAZY
scripts/config -e PREEMPT_RT
scripts/config -e PREEMPT_COUNT
scripts/config -e PREEMPTION
scripts/config -d PREEMPT_DYNAMIC
%endif


mv .config config

%build
BuildKernel() {

    Target=$1
    Arch=x86_64
    ExtraVer="-%{release}.${Target}"

    rm -f localversion-rt localversion-xanmod

    perl -p -i -e "s/^EXTRAVERSION.*/EXTRAVERSION = ${ExtraVer}/" Makefile

    make O=${Target} -s mrproper
    cp config ${Target}/.config

    make O=${Target} -s ARCH=${Arch} olddefconfig
    make O=${Target} -s ARCH=${Arch} CONFIG_DEBUG_SECTION_MISMATCH=y %{?_smp_mflags} %{?sparse_mflags}
}

BuildKernel %{ktarget}

%install

InstallKernel() {

    Target=$1
    Kversion=$2
    Arch=x86_64
    KernelDir=%{buildroot}/usr/lib/kernel
    DevDir=%{buildroot}/usr/lib/modules/${Kversion}/build

    mkdir   -p ${KernelDir}
    install -m 644 ${Target}/.config    ${KernelDir}/config-${Kversion}
    install -m 644 ${Target}/System.map ${KernelDir}/System.map-${Kversion}
    install -m 644 ${Target}/vmlinux    ${KernelDir}/vmlinux-${Kversion}
    install -m 644 %{SOURCE2}           ${KernelDir}/cmdline-${Kversion}
    cp  ${Target}/arch/x86/boot/bzImage ${KernelDir}/org.clearlinux.${Target}.%{version}-%{release}
    chmod 755 ${KernelDir}/org.clearlinux.${Target}.%{version}-%{release}

    mkdir -p %{buildroot}/usr/lib/modules
    make O=${Target} -s ARCH=${Arch} INSTALL_MOD_PATH=%{buildroot}/usr modules_install

    rm -f %{buildroot}/usr/lib/modules/${Kversion}/build
    rm -f %{buildroot}/usr/lib/modules/${Kversion}/source

    mkdir -p ${DevDir}
    find . -type f -a '(' -name 'Makefile*' -o -name 'Kbuild*' -o -name 'Kconfig*' ')' -exec cp -t ${DevDir} --parents -pr {} +
    find . -type f -a '(' -name '*.sh' -o -name '*.pl' ')' -exec cp -t ${DevDir} --parents -pr {} +
    cp -t ${DevDir} -pr ${Target}/{Module.symvers,tools}
    ln -s ../../../kernel/config-${Kversion} ${DevDir}/.config
    ln -s ../../../kernel/System.map-${Kversion} ${DevDir}/System.map
    cp -t ${DevDir} --parents -pr arch/x86/include
    cp -t ${DevDir}/arch/x86/include -pr ${Target}/arch/x86/include/*
    cp -t ${DevDir}/include -pr include/*
    cp -t ${DevDir}/include -pr ${Target}/include/*
    cp -t ${DevDir} --parents -pr scripts/*
    cp -t ${DevDir}/scripts -pr ${Target}/scripts/*
    find  ${DevDir}/scripts -type f -name '*.[cho]' -exec rm -v {} +
    find  ${DevDir} -type f -name '*.cmd' -exec rm -v {} +
    # Cleanup any dangling links
    find ${DevDir} -type l -follow -exec rm -v {} +

    # Kernel default target link
    ln -s org.clearlinux.${Target}.%{version}-%{release} %{buildroot}/usr/lib/kernel/default-${Target}
}

# cpio file for keyboard drivers
createCPIO() {

    Target=$1
    Kversion=$2
    KernelDir=%{buildroot}/usr/lib/kernel
    ModDir=/usr/lib/modules/${Kversion}

    mkdir -p cpiofile${ModDir}/kernel/drivers/input/{serio,keyboard}
    mkdir -p cpiofile${ModDir}/kernel/drivers/hid
    cp %{buildroot}${ModDir}/kernel/drivers/input/serio/i8042.ko      cpiofile${ModDir}/kernel/drivers/input/serio
    cp %{buildroot}${ModDir}/kernel/drivers/input/serio/libps2.ko     cpiofile${ModDir}/kernel/drivers/input/serio
    cp %{buildroot}${ModDir}/kernel/drivers/input/keyboard/atkbd.ko   cpiofile${ModDir}/kernel/drivers/input/keyboard
    cp %{buildroot}${ModDir}/kernel/drivers/hid/hid-logitech-dj.ko    cpiofile${ModDir}/kernel/drivers/hid
    cp %{buildroot}${ModDir}/kernel/drivers/hid/hid-logitech-hidpp.ko cpiofile${ModDir}/kernel/drivers/hid
    cp %{buildroot}${ModDir}/kernel/drivers/hid/hid-apple.ko          cpiofile${ModDir}/kernel/drivers/hid
    cp %{buildroot}${ModDir}/modules.order   cpiofile${ModDir}
    cp %{buildroot}${ModDir}/modules.builtin cpiofile${ModDir}

    depmod -b cpiofile/usr ${Kversion}

    (
      cd cpiofile
      find . | cpio --create --format=newc \
        | xz --check=crc32 --lzma2=dict=512KiB > ${KernelDir}/initrd-org.clearlinux.${Target}.%{version}-%{release}
    )
}

InstallKernel %{ktarget} %{kversion}

createCPIO %{ktarget} %{kversion}

rm -rf %{buildroot}/usr/lib/firmware

mkdir -p %{buildroot}/usr/share/package-licenses/linux-xmrt
cp COPYING %{buildroot}/usr/share/package-licenses/linux-xmrt/COPYING
cp -a LICENSES/* %{buildroot}/usr/share/package-licenses/linux-xmrt

%postun
rm -fr /var/lib/dkms/*/kernel-%{kversion}-x86_64
rm -fr /var/lib/dkms/*/*/%{kversion}

%files
%dir /usr/lib/kernel
%dir /usr/lib/modules/%{kversion}
/usr/lib/kernel/config-%{kversion}
/usr/lib/kernel/cmdline-%{kversion}
/usr/lib/kernel/org.clearlinux.%{ktarget}.%{version}-%{release}
/usr/lib/kernel/default-%{ktarget}
/usr/lib/modules/%{kversion}/kernel
/usr/lib/modules/%{kversion}/modules.*

%files extra
%dir /usr/lib/kernel
/usr/lib/kernel/System.map-%{kversion}
/usr/lib/kernel/vmlinux-%{kversion}

%files license
%defattr(0644,root,root,0755)
/usr/share/package-licenses/linux-xmrt

%files cpio
/usr/lib/kernel/initrd-org.clearlinux.%{ktarget}.%{version}-%{release}

%files dev
%defattr(-,root,root)
/usr/lib/modules/%{kversion}/build

```

</details>

**Comment by marioroy** [comment]
I disabled the RCU section in the spec file. That resolved both issues. Also, no frame drops playing a video in Chrome without `chrt`.
```text
%if 0
scripts/config -e RCU_EXPERT
scripts/config -d FORCE_TASKS_RCU
scripts/config -d FORCE_TASKS_RUDE_RCU
scripts/config -d FORCE_TASKS_TRACE_RCU
scripts/config --set-val RCU_FANOUT 64
scripts/config --set-val RCU_FANOUT_LEAF 16
scripts/config -e RCU_BOOST
scripts/config --set-val RCU_BOOST_PRIO 99
scripts/config --set-val RCU_BOOST_DELAY 331
scripts/config -d RCU_EXP_KTHREAD
scripts/config -e RCU_NOCB_CPU
scripts/config -e RCU_NOCB_CPU_DEFAULT_ALL
scripts/config -d RCU_NOCB_CPU_CB_BOOST
scripts/config -d TASKS_TRACE_RCU_READ_MB
scripts/config -d RCU_LAZY
%endif

```

**Comment by marioroy** [comment]
On Clear Linux, I'm unable to use the XanMod config e.g. config_x86-64-v3. The system does not boot. So, I built the XanMod kernels using the Clear Linux config.

New plan: The spec files now apply XanMod default settings for tick rate, timers, rcu, and preempt/-rt.

**Comment by marioroy** [comment]
On Clear Linux, running `primesieve 3e12` does not affect desktop applications. The Clear config has the mentioned knobs enabled. So does XanMod's configs. Could the following be broken in XanMod?

```text
config SCHED_AUTOGROUP
        bool "Automatic process group scheduling"
        select CGROUPS
        select CGROUP_SCHED
        select FAIR_GROUP_SCHED
        help
          This option optimizes the scheduler for common desktop workloads by
          automatically creating and populating task groups.  This separation
          of workloads isolates aggressive CPU burners (like build jobs) from
          desktop applications.  Task group autogeneration is currently based
          upon task session.
        
config SCHED_AUTOGROUP_DEFAULT_ENABLED 
        bool "Enable automatic process group scheduling feature"
        default y
        depends on SCHED_AUTOGROUP
        help
          If set, automatic process group scheduling will be enabled per
          default but can be disabled through passing autogroup=0 on the
          kernel commandline during boot or a value of 0 via the file
          proc/sys/kernel/sched_autogroup_enabled.
          
          If unsure say Y.
```

**Comment by marioroy** [comment]
The `sched_autogroup_enabled` is not enabled by default. No problem. I will add a kernel parameter for the XanMod on Clear Linux project.
```text
# echo 1 >/proc/sys/kernel/sched_autogroup_enabled
```
That resolved the issue 100% for Chrome freezing while moving the window and primesieve running in the background consuming all CPU cores. I have one more thing to try. Playing a video is still dropping frames but much less than before.

**Comment by marioroy** [comment]
`PREEMPT_DYNAMIC` was missing. I can now leave preempt and rcu boost enabled by default. No pause moving the Chrome window or frame drops watching a video.

```text
# Enable preempt.
%if 1
scripts/config -e PREEMPT_BUILD 
scripts/config -d PREEMPT_NONE
scripts/config -d PREEMPT_VOLUNTARY
scripts/config -e PREEMPT 
scripts/config -e PREEMPT_COUNT
scripts/config -e PREEMPTION
scripts/config -e PREEMPT_DYNAMIC
%endif

# Enable RCU boost (requires preempt).
%if 1
scripts/config -e RT_MUTEXES
scripts/config -e PREEMPT_RCU
scripts/config -e RCU_EXPERT
scripts/config -e RCU_BOOST
scripts/config --set-val RCU_BOOST_DELAY 500
scripts/config -d RCU_EXP_KTHREAD
%endif
```

**Comment by marioroy** [comment]
I could crash the prior rt kernel and was unsure about inclusion in the xanmod-on-clear project. However, the newly released `6.1.54-rt15-xanmod1` kernel is passing. I'm unable to crash it under load and moving the Chrome window in a circle around the desktop. No frame drops as well (made possible with autogroup).

The spec files are based on Clear's kernel and using Clear's config file. The mini-overrides is all that I'm doing. Folks may disable if so desired for better performance. 
```text
cp %{SOURCE1} .config

# Enable preempt.
%if 1
scripts/config -e PREEMPT_BUILD 
scripts/config -d PREEMPT_NONE
scripts/config -d PREEMPT_VOLUNTARY
scripts/config -e PREEMPT 
scripts/config -e PREEMPT_COUNT
scripts/config -e PREEMPTION
scripts/config -e PREEMPT_DYNAMIC
%endif

# Enable RCU boost (requires preempt).
%if 1
scripts/config -e RT_MUTEXES
scripts/config -e PREEMPT_RCU
scripts/config -e RCU_EXPERT
scripts/config -e RCU_BOOST
scripts/config --set-val RCU_BOOST_DELAY 500
scripts/config -d RCU_EXP_KTHREAD
%endif

mv .config config
```

And likewise for the RT kernel.
```text
cp %{SOURCE1} .config

# Enable preempt-rt.
%if 1
scripts/config -e PREEMPT_BUILD
scripts/config -d PREEMPT_NONE 
scripts/config -d PREEMPT_VOLUNTARY
scripts/config -d PREEMPT
scripts/config -e PREEMPT_RT
scripts/config -e PREEMPT_COUNT
scripts/config -e PREEMPTION
%endif

# Enable RCU boost (requires preempt-rt).
%if 1
scripts/config -e RT_MUTEXES
scripts/config -e PREEMPT_RCU
scripts/config -e RCU_EXPERT
scripts/config -e RCU_BOOST
scripts/config --set-val RCU_BOOST_DELAY 500
scripts/config -d RCU_EXP_KTHREAD
%endif

mv .config config
```

This is a personal project because I like the XanMod kernels and curiosity running a RT kernel.

**Comment by marioroy** [comment]
I will do five flavors.
```text
1. xmedge-desktop (preempt and rcu_boost enabled)
2. xmedge-server
3. xmlts-desktop (preempt and rcu_boost enabled)
4. xmlts-server
5. xmrt (preempt-rt and rcu_boost enabled)
```

**Comment by marioroy** [comment]
Benchmark results on Clear Linux.

https://gist.github.com/marioroy/7a6384286f367e53758072962ad36c7f

I will have two rt preempt flavors; one configured with PREEMPT_RT, the other PREEMPT_DYNAMIC.

**Comment by marioroy** [comment]
The project is completed.

https://community.clearlinux.org/t/run-xanmod-kernels-on-clear-linux/9161

**Comment by marioroy** [comment]
The solution to the issue describe above is modifying `latency_ns` and `min_granularity_ns` to the values below. With the system under load (primesieve 4e12), I can move the Chrome window around the desktop without stalling or interruption.  Also passing, is playing a video 1440p 60fps (no frame drops) while running `primesieve 4e12`. I have sched autogroup disabled as well.

```bash
LATENCY_NS_FILE="/sys/kernel/debug/sched/latency_ns"

   Clear LTS: 24000000
  XanMod LTS:  6000000
    Modified: 12000000

MIN_GRANULARITY_NS_FILE="/sys/kernel/debug/sched/min_granularity_ns"

   Clear LTS: 3000000
  XanMod LTS:  750000
    Modified: 1500000

WAKEUP_GRANULARITY_NS_FILE="/sys/kernel/debug/sched/wakeup_granularity_ns"

   Clear LTS: 4000000
  XanMod LTS: 3000000

MIGRATION_COST_NS_FILE="/sys/kernel/debug/sched/migration_cost_ns"

   Clear LTS: 500000
  XanMod LTS: 500000

BANDWIDTH_SIZE_US_FILE="/proc/sys/kernel/sched_cfs_bandwidth_slice_us"

   Clear LTS: 5000
  XanMod LTS: 5000

NR_MIGRATE_FILE="/sys/kernel/debug/sched/nr_migrate"

   Clear LTS: 32
  XanMod LTS: 32
```

@xanmod ^^ All tests pass. So now, I will make a patch file to change `latency_ns` and `min_granularity_ns` to the modified values. This resolves the window movement stalling or frames dropping while running `primesieve 4e12`.

Video (select quality 1440p):
Big Buck Bunny 60fps 4K - Official Blender Foundation Short Film
https://www.youtube.com/watch?v=aqz-KE-bpKQ
